### PR TITLE
explicitly close socat socket

### LINKED ISSUE #11621: fix socket timeout exceptions on failed stream close on kubernetes
The underlying issue for https://github.com/airbytehq/oncall/issues/140 is possible to reproduce across a portion of syncs run in load testing.

### PR DESCRIPTION
**Pull request by git-phu**
## What
closes #11621

Currently the way we handle sending messages from the source pod to the orchestrator pod is:
- The orchestrator loops forever trying to read messages from the source until it detects that the source pod has terminated
- The source pod uses socat to transfer a stream of data to the orchestrator.
    - Once the stream's EOF is reached, socat shuts down its write stream.
    - After a 0.5 second timeout, socat closes the connection to the orchestrator as well and the source pod terminates

We can see this in a sample of the socat logs.
```
2022/05/05 02:03:29 socat[9] I transferred 8192 bytes from 0 to 5
2022/05/05 02:03:29 socat[9] I transferred 5306 bytes from 0 to 5
2022/05/05 02:03:29 socat[9] N socket 1 (fd 0) is at EOF
2022/05/05 02:03:29 socat[9] I shutdown(5, 1)
2022/05/05 02:03:30 socat[9] I poll timed out (no data within 0.500000 seconds)
2022/05/05 02:03:30 socat[9] I shutdown(5, 2)
2022/05/05 02:03:30 socat[9] N exiting with status 0
```

The problem with this approach is that we have seen cases where it appears that the source pod has already terminated before the orchestrator is able to fully process all the incoming data. This leaves the network socket in a bad state and the orchestrator pod simply hangs.  https://github.com/airbytehq/airbyte/issues/11621#issuecomment-1118828858 for more details on how this might happen.

A more reliable way to handle this interaction should be:
- The orchestrator loops forever trying to read messages from the source
- The source pod uses socat to transfer a stream of data to the orchestrator.
    - Once the stream's EOF is reached, socat shuts down its write stream.
- **Once the orchestrator receives the EOF from socat, the orchestrator should initiate shutting down the network socket**
    - after socat receives the signal to shut down the read stream from the orchestrator, the network socket is closed and the source pod returns.

This new flow should avoid the problems arising from the source initiating the shutdown.  This is also how socat is intended to work.
http://www.dest-unreach.org/socat/doc/socat.html

> When one of the streams effectively reaches EOF, the closing phase begins. Socat transfers the EOF condition to the other stream, i.e. tries to shutdown only its write stream, giving it a chance to terminate gracefully. For a defined time socat continues to transfer data in the other direction, but then closes all remaining channels and terminates.

## How
- Change [DefaultReplicatonWorker#getReplicationRunnable](https://github.com/airbytehq/airbyte/blob/49810f8c139297c9261b5be6ce93cbd124c69f03/airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java#L271) so that when there are no more messages from the source, the orchestrator initiate closing the network connection.
- Increase the timeout on socat to 60 seconds (from the default 0.5 seconds) to give the orchestrator time to receive all messages and then initiate shutdown of the network connection.

With this change the socat logs now look like
```
2022/05/05 16:09:02 socat[8] N reading from and writing to stdio
2022/05/05 16:09:02 socat[8] N opening connection to AF=2 172.25.37.3:9877
2022/05/05 16:09:02 socat[8] N successfully connected from local address AF=2 172.25.32.134:38950
2022/05/05 16:09:02 socat[8] N starting data transfer loop with FDs [0,1] and [5,5]
2022/05/05 16:14:17 socat[8] N socket 1 (fd 0) is at EOF
2022/05/05 16:14:19 socat[8] N socket 2 (fd 5) is at EOF
2022/05/05 16:14:19 socat[8] N exiting with status 0
```
(note that socket 2 also has an EOF now)

### CODE REVIEW

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java [hunk 1]
```diff
@@ -300,6 +300,13 @@ private static Runnable getReplicationRunnable(final AirbyteSource source,
             if (recordsRead % 1000 == 0) {
               LOGGER.info("Records read: {} ({})", recordsRead, FileUtils.byteCountToDisplaySize(messageTracker.getTotalBytesEmitted()));
             }
+          } else {
+            LOGGER.info("Source has no more messages, closing connection.");
+            try {
+              source.close();
+            } catch (final Exception e) {
+              throw new SourceException("Source cannot be stopped!", e);
```
Comment (supertopher) [inline airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java:308]: I love this exception.  it sounds like Godzilla is coming from the water and cannot be contained

**Review by supertopher — APPROVED** [review]
looks great Peter

File: airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java [hunk 2]
```diff
@@ -133,7 +134,7 @@ void test() throws Exception {
     verify(destination).start(destinationConfig, jobRoot);
     verify(destination).accept(RECORD_MESSAGE1);
     verify(destination).accept(RECORD_MESSAGE2);
-    verify(source).close();
+    verify(source, atLeastOnce()).close();
```
Comment (xiaohansong) [inline airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java:137]: I might not have understood this completely - in which scenario the source might be closed for more than once?

Comment (git-phu) [reply airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java:137]: ah good point.  I made a brief mention of this in the commit message [here](https://github.com/airbytehq/airbyte/pull/12632/commits/9831d347fe5c9f2b28bdea951eddd683eece23e2) but basically the source implements the AutoCloseable interface, and I am also explicitly calling close with this change, so close ends up being called twice

**Review by xiaohansong — APPROVED** [review]

**Comment by git-phu** [comment]
did a final verification in a dev env that everything works as expected (discovers, syncs, etc.)

**Pull request merged**